### ISSUE TITLE
Make it possible for find() to strip loki metadata

### ISSUE THREAD

**Issue by alexandernst** [techfort/LokiJS#346]
Currently a `find()` returns the data itself and some loki-related metadata. While useful for debugging purposes, this can't be expected to happen in a production environment. It also adds extra difficulties when trying to write unit tests against inserted data.

**Comment by alexandernst** [comment]
@techfort I'll be working on this, implemented as [Mongo's projections functionality](https://docs.mongodb.org/v3.0/reference/method/db.collection.find/). I'm looking at the code, but I can't decide where to put the `projections` argument in the `Resultset` class.

`Resultset` is currently used by a few public methods (especially `find`). Adding the `projections` argument there is easy (`find(query, projection)`), but adding it in the private `Resultset` class is a little bit more tricky.

Currently `Resultset` is defined as `(collection, queryObj, queryFunc, firstOnly)`. Do you think it would be better to change it as `(collection, queryObj, queryFunc, firstOnly, projection)` or `(collection, queryObj, projection, queryFunc, firstOnly)`?  (or maybe any other way).

The second one would require current private methods that call `Resultset` and use `queryFunc` to call it with one more argument (example: `Resultset(this, query, null)` would become `Resultset(this, query, null, null)`).

**Comment by alexandernst** [comment]
@techfort Another thing that must be discussed. The `projection` can't have both `include` and `exclude` keys. They all must be `all included except...` or `all excluded except...`.
If the developer who is using the library makes the error of passing both types, what should be the behavior?

* completely ignore the entire `projection` argument and return everything?
* return `none` of some kind? (let it be `undefined`, `null`, empty array, etc...)
* throw some kind of an error?

**Comment by techfort** [comment]
@alexandernst not much time to ponder everything in detail but i would rather `(collection, queryObj, queryFunc, firstOnly, projection)` however @obeliskos may have stronger feelings has he developed Resultset in the first place. As for the include/exclude issue, i would possibly throw an error of the "make your mind up man" type.

**Comment by alexandernst** [comment]
@obeliskos Can you enlighten me up on this, please? How would you feel better to implement this?

**Comment by obeliskos** [comment]
I believe in the past/currently you would just do a map of results, but sure, I guess you could do that. 

Ideally i would think we would clean up those params by putting then in a single options object... not sure how many people are instantiating their own resultset classes so i'll let you guys decide if that is 'breaking' change.  You'd still need to track down all internal resultset instantiations and update their params.

As for implementation, resultset find() code path is different for non-chained coll.find() vs chained finds.  Figuring out where to inject your logic in non-chained path where we just return object array might be tricky.

I'd recommend leaving logic as-is in resultset.find and adding a projection() function to resultset similar to map().  The user can either chain project or you can detect when a user is attempting a collection.find() projection and alter the code path to resolve it by instantiating a 'chained' resultset (instead of calling it like a function) and chain the find(), projection(), and data()?  I guess your projection would return document array or would it create its own collection copy?

**Comment by VladimirTechMan** [comment]
@obeliskos @techfort @alexandernst If I may add my two cents here, I would support introducing the `options` argument at this point, making it the final argument so we would have `(collection, queryObj, queryFunc, firstOnly, options)`. Then wrapping `projections` inside `options`.

*I'd love to wrap the preceding options, expect `collection` (and maybe `queryObj` / `queryFunc`), into `options`, but that'd break backward compatibility. Thus, let's postpone it till 2.0.*

@alexandernst Let me know if you have any concerns about it. The extra code to handle the `options` argument is minimal. It would not be too much of extra work and extra size for the library. If you need any help working on that part, let us know.

**Comment by alexandernst** [comment]
First of all, sorry for delaying the conversation flow. I wasn't available yesterday, due to a temporary project swap.

Is `Resultset` even a publicly available class? If it's not explicitly public, I'm all in for just wrapping everything in an object, like @obeliskos said, without waiting for the 2.0. People should know that using internal classes/methods is something that they shouldn't do.

@obeliskos Ah, thanks for the tip. I should look at that carefully (the different path codes). And indeed, the `projection()` chain idea is something that I might use, it sounds like it could be reused in the future.

Summing up, we must decide if `Resultset` is public. If not, I'll make a PR to change the arguments to a single object, then I'll start working on this issue itself.

**Comment by techfort** [comment]
@alexandernst the idea of the argument being an options argument (as suggested by @VladimirTechMan ) is good, extensible, and the rationale why we moved to this approach for plain collections too (it used to be a 3-4 args constructor), so i'm cool with that.
@obeliskos have you any objections to exposing `Resultset`? seems like many devs want underlying classes exposed, so I don't see why not, but maybe you've a good reason to keep it private.

**Comment by obeliskos** [comment]
Not sure of the benefit of exposing resultset.  It requires passing in collection ref to constructor, which is done on collection.chain() which returns a reference to resultset instantiated with collection ref and null for both queryobj and queryfunc.  

So isn't collection.chain() the best way to get a new Resultset object bound to a specific collection (context)?  

Also at each step in a query chain you can use the resultset return value to introspect rather than immediately calling data if you want to examine internals at each step of chain.  At any point you can take an existing resultset and copy() aka branch() it also.

**Comment by alexandernst** [comment]
I'm also -1 for exposing publicly `Resultset`. Internal data structure/classes/methods shouldn't be exposed and shouldn't be allowed to be used by third parties. Only the public API should be kept intact.

**Comment by techfort** [comment]
@obeliskos i think i get your point . i do find it difficult to picture a scenario in which someone would create an instance of resultset without this coming from a query.  In that respect it's probably going to cause more problems than benefits to have a public `Resultset`. And as you say, the resultset object can be easily inspected, so that is not an issue either. @alexandernst @VladimirTechMan @obeliskos i must apologize because reading the posts above I realize nobody really asked for Resultset to be exposed, i misunderstood in the haste of skimming.
Are we all agreed on an options argument? as soon as we break backward compatibility I would also migrate `firstOnly` into the `options`, and potentially even `queryFunc` but we can discuss that when we cross that bridge.

**Comment by alexandernst** [comment]
@techfort As far as modifying `Resultset`, there won't be any breaking changes (as `Resultset` isn't really public). So we can "break" as much stuff as we want there, without any risks.

I think it's clear now, so let me make some initial PRs for that.

**Comment by alexandernst** [comment]
@techfort @obeliskos @VladimirTechMan I done the refactoring we spoke about yesterday. Anyways, I think I may have found a bigger problem. When `find` is used with `chain`, each consecutive chained function expects a full resultset.

Example:

    chain().find().update() <--- update expects find to return a full ResultSet data

If we remove the `$meta` and/or `$loki` keys from the object we return from `find()`, `update()` won't work (as it will think that the documents haven't been inserted yet, as they miss the metadata info).

**Comment by sondreb** [comment]
I'm on LokiJS 1.3.19 and .find does not return $loki and meta, I'm trying to find a solution to actually return those properties using the find operator.

coll.data <- returns $loki and meta
coll.find({}) <- does not return $loki and meta

Any ideas?

**Comment by dmccuskey** [comment]
how about if you put this functionality in the `data()` method ? i was already expecting it to be there since it terminates the query chain and, due to the method name, it already seems like that's what it's going to do - give me back my data.

if i'm chaining things together, i'm ok with some "magic" along the way to help things along (a la Promises, meta, $loki), but i think there should be some way to get my "pure" data back.

my use case is sending results back to front-end for display/pagination. for now i'm using the following. *(pasted here for others who may want a quick copy/paste solution)*

```
export function stripResultsMetadata( results ) {
	const records = []
	for (var idx = 0; idx < results.length; idx++) {
		const loki_rec = results[ idx ]
		const clean_rec = Object.assign({}, loki_rec)
		delete clean_rec['meta']
		delete clean_rec['$loki']
		records.push( clean_rec )
	}
	return records
}
```

ps, thanks for the great work on lokijs !
